File: cookiesync/cookie/crypto.py
"""Chrome macOS cookie encryption and decryption (the ``v10`` scheme).

key   = PBKDF2-HMAC-SHA1(safe_storage_password, b"saltysalt", 1003, dklen=16)
value = AES-128-CBC(key, iv=16x 0x20) over the ciphertext, PKCS7-(un)padded, with a
        32-byte SHA256(host_key) domain-hash prefix Chrome v24+ prepends — verified on
        decrypt (a hash mismatch is a wrong key, not garbage), and committed on encrypt
        against the exact stored ``host_key`` (leading dot included).

``v20`` (app-bound) values cannot be (de)crypted with the Safe Storage key and are rejected.
"""

from __future__ import annotations

import hashlib

from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes

from cookiesync.cookie.models import AesKey, HostKey, SafeStorageKey

SALT = b"saltysalt"
ITERATIONS = 1003
KEY_LENGTH = 16
IV = b"\x20" * 16


class DecryptError(Exception):
    """A cookie value could not be decrypted (v20, malformed, or wrong key)."""


def derive_key(password: SafeStorageKey) -> AesKey:
    """Derive the 16-byte AES key from the raw 'Safe Storage' password."""
    return AesKey(hashlib.pbkdf2_hmac("sha1", password.encode("utf-8"), SALT, ITERATIONS, dklen=KEY_LENGTH))


def pkcs7_pad(data: bytes) -> bytes:
    pad = 16 - len(data) % 16
    return data + bytes([pad]) * pad


def pkcs7_unpad(data: bytes) -> bytes:
    if not data:
        raise DecryptError("empty plaintext")
    pad = data[-1]
    if pad < 1 or pad > 16 or pad > len(data):
        raise DecryptError(f"bad PKCS7 padding length {pad}")
    if data[-pad:] != bytes([pad]) * pad:
        raise DecryptError("inconsistent PKCS7 padding")
    return data[:-pad]


def domain_hash(host_key: HostKey) -> bytes:
    return hashlib.sha256(host_key.encode("utf-8")).digest()


def decrypt_value(encrypted: bytes, key: AesKey, host_key: HostKey) -> str:
    """Decrypt one Chrome cookie ``encrypted_value``. Raise ``DecryptError`` on failure."""
    if not encrypted:
        return ""
    match encrypted[:3]:
        case b"v20":
            raise DecryptError("v20 app-bound cookie (not decryptable with the Safe Storage key)")
        case b"v10":
            ciphertext = encrypted[3:]
        case _:
            try:
                return encrypted.decode("utf-8")
            except UnicodeDecodeError as exc:
                raise DecryptError("unrecognized cookie encoding") from exc
    if not ciphertext or len(ciphertext) % 16 != 0:
        raise DecryptError("ciphertext is not a positive multiple of the block size")
    decryptor = Cipher(algorithms.AES(key), modes.CBC(IV)).decryptor()
    plain = pkcs7_unpad(decryptor.update(ciphertext) + decryptor.finalize())
    if len(plain) < 32 or plain[:32] not in {domain_hash(host_key), domain_hash(HostKey(host_key.lstrip(".")))}:
        raise DecryptError("domain-hash prefix mismatch (wrong key)")
    try:
        return plain[32:].decode("utf-8")
    except UnicodeDecodeError as exc:
        raise DecryptError("decrypted value is not valid UTF-8 (likely wrong key)") from exc


def encrypt_value(plaintext: str, key: AesKey, host_key: HostKey) -> bytes:
    """Encrypt one cookie value into Chrome's ``v10`` blob, committing to the exact ``host_key``."""
    encryptor = Cipher(algorithms.AES(key), modes.CBC(IV)).encryptor()
    block = pkcs7_pad(domain_hash(host_key) + plaintext.encode("utf-8"))
    return b"v10" + encryptor.update(block) + encryptor.finalize()

File: internal/cookie/crypto.go
package cookie

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha1" //nolint:gosec // Chrome's Safe Storage KDF is fixed at PBKDF2-HMAC-SHA1; parity, not a security choice.
	"crypto/sha256"
	"errors"
	"fmt"
	"strings"
	"unicode/utf8"

	"golang.org/x/crypto/pbkdf2"
)

// Chrome macOS "Safe Storage" v10 crypto parameters. These are fixed by Chrome and
// must match byte-for-byte:
//
//	key   = PBKDF2-HMAC-SHA1(safe_storage_password, "saltysalt", 1003, dklen=16)
//	value = AES-128-CBC(key, iv=16x 0x20) over the ciphertext, PKCS7-(un)padded, with
//	        a 32-byte SHA256(host_key) domain-hash prefix Chrome v24+ prepends.
const (
	iterations = 1003
	keyLength  = 16
)

var (
	salt = []byte("saltysalt")
	iv   = bytes.Repeat([]byte{0x20}, 16)
)

// ErrV20 marks a v20 (app-bound) cookie value, which cannot be decrypted with the
// Safe Storage key. The pipeline counts these separately from other decrypt
// failures, so callers branch on it with errors.Is(err, ErrV20).
var ErrV20 = errors.New("v20 app-bound cookie (not decryptable with the Safe Storage key)")

// DecryptError reports that a cookie value could not be decrypted: a v20 app-bound
// blob, a malformed ciphertext, or a wrong key. Its message mirrors the Python
// implementation byte-for-byte.
type DecryptError struct {
	Msg string
	Err error
}

func (e *DecryptError) Error() string { return e.Msg }

func (e *DecryptError) Unwrap() error { return e.Err }

// DeriveKey derives the 16-byte AES key from the raw "Safe Storage" password.
func DeriveKey(password SafeStorageKey) AesKey {
	return AesKey(pbkdf2.Key([]byte(password), salt, iterations, keyLength, sha1.New))
}

func pkcs7Pad(data []byte) []byte {
	pad := 16 - len(data)%16
	return append(data, bytes.Repeat([]byte{byte(pad)}, pad)...)
}

func pkcs7Unpad(data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, &DecryptError{Msg: "empty plaintext"}
	}
	pad := int(data[len(data)-1])
	if pad < 1 || pad > 16 || pad > len(data) {
		return nil, &DecryptError{Msg: fmt.Sprintf("bad PKCS7 padding length %d", pad)}
	}
	if !bytes.Equal(data[len(data)-pad:], bytes.Repeat([]byte{byte(pad)}, pad)) {
		return nil, &DecryptError{Msg: "inconsistent PKCS7 padding"}
	}
	return data[:len(data)-pad], nil
}

func domainHash(hostKey HostKey) []byte {
	sum := sha256.Sum256([]byte(hostKey))
	return sum[:]
}

// DecryptValue decrypts one canonical Chrome v10 encrypted_value blob. It returns a
// *DecryptError on any failure; a v20 blob unwraps to ErrV20.
func DecryptValue(encrypted []byte, key AesKey, hostKey HostKey) (string, error) {
	var ciphertext []byte
	switch {
	case bytes.HasPrefix(encrypted, []byte("v20")):
		return "", &DecryptError{Msg: ErrV20.Error(), Err: ErrV20}
	case bytes.HasPrefix(encrypted, []byte("v10")):
		ciphertext = encrypted[3:]
	default:
		return "", &DecryptError{Msg: "unrecognized cookie encoding"}
	}

	if len(ciphertext) == 0 || len(ciphertext)%aes.BlockSize != 0 {
		return "", &DecryptError{Msg: "ciphertext is not a positive multiple of the block size"}
	}

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", fmt.Errorf("aes cipher: %w", err)
	}
	plain := make([]byte, len(ciphertext))
	cipher.NewCBCDecrypter(block, iv).CryptBlocks(plain, ciphertext)

	plain, err = pkcs7Unpad(plain)
	if err != nil {
		return "", err
	}

	if len(plain) < 32 || !domainHashMatches(plain[:32], hostKey) {
		return "", &DecryptError{Msg: "domain-hash prefix mismatch (wrong key)"}
	}

	value := plain[32:]
	if !utf8.Valid(value) {
		return "", &DecryptError{Msg: "decrypted value is not valid UTF-8 (likely wrong key)"}
	}
	return string(value), nil
}

// domainHashMatches dual-accepts the SHA256 of the host_key as stored and with its
// leading dot stripped, matching how Chrome commits either form.
func domainHashMatches(prefix []byte, hostKey HostKey) bool {
	return bytes.Equal(prefix, domainHash(hostKey)) ||
		bytes.Equal(prefix, domainHash(HostKey(strings.TrimLeft(string(hostKey), "."))))
}

// EncryptValue encrypts one cookie value into Chrome's v10 blob, committing to the
// exact host_key (leading dot included) via the 32-byte domain-hash prefix.
func EncryptValue(plaintext string, key AesKey, hostKey HostKey) ([]byte, error) {
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("aes cipher: %w", err)
	}
	padded := pkcs7Pad(append(domainHash(hostKey), plaintext...))
	out := make([]byte, len(padded))
	cipher.NewCBCEncrypter(block, iv).CryptBlocks(out, padded)
	return append([]byte("v10"), out...), nil
}
